### ISSUE TITLE
Support for app.postDeploy() for post-deploy functions

### ISSUE THREAD

**Issue by KingOfPoptart** [awslabs/aws-cdk#2849]
A post-deploy hook would be handy - sometimes its useful to have something that runs after a successful `cdk deploy`.  Otherwise, CDK will force end users have CDK run in wrapper scripts

What would also be handy is if during this `postDeploy()` functionality, if cloudFormation resources were no longer tokens, but were fully fleshed out and available

Something like the following would be convenient - 

```node
import cdk = require('@aws-cdk/cdk');

function runAfterCompletion(NameOfResource: string){
    notifyMyTeam(`mySuperCoolStack has deployed ${NameOfResource}`);
}


const app = new cdk.App();
const myAsg = new mySuperCoolAsgStack(app, "VPC", {props: "props"});
app.run()

app.postDeploy(runAfterCompletion, myAsg.GetAsg.autoScalingGroupName);
```

**Comment by bgdnlp** [comment]
Very useful for working around missing CloudFormation features.

Example use case: As far as I can tell, at the moment it's not possible to tell CloudFormation to set a Cognito user pool to send a verification link in email instead of a code ([VerificationMessageTemplateType](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_VerificationMessageTemplateType.html#CognitoUserPools-Type-VerificationMessageTemplateType-DefaultEmailOption) in the API). With a post-deploy hook it would at least be possible to set it after the pool is created.

Pre-deploy might also be useful. Also, a separate hook for each stack if possible, and one after everything was deployed.

**Comment by eladb** [comment]
I am wondering if custom resources would be sufficient for these use cases...

**Comment by bgdnlp** [comment]
Custom resources can't replace hooks completely. Leaving aside the ease of use, the main difference is where the code is executed.

Simple example. Use the CDK as part of CD/CI pipeline that runs on a Jenkins server. The CDK deals with EC2, so the Jenkins role only has access to EC2. With a hook, that's enough. With a custom resource, Jenkins now needs access to set up and run arbitrary Lambda functions, along with the roles attached to those functions.

Another example, a simple notification before and after deploy. To Slack, email, whatever. With a hook it's just a local script. Custom resources need more setup.

Many use cases can be worked around with custom resources. But it's clumsy and it complicates things considerably. Especially in bigger organisations, where access has to be vetted by a different team.

**Comment by bgdnlp** [comment]
From Gitter, another use case for local hooks, as opposed to custom resources: "For instance, Account A stack has an iam role that Account B's bucket would like to give access to".

This could be easily solved with a post deploy hook on AccountAStack which gets the IAM role from Account A, then passes it to AccountBStack as a context variable, or in a pre-deploy hook.

To clarify, the kind of hooks I'm thinking about are the kind that [Sceptre implemented](https://sceptre.cloudreach.com/2.1.5/docs/hooks.html). Pre/post each stack's deploy/delete. Sceptre follows the same basic idea that the CDK has, it provides a cli for building (paired with Troposphere), deploying and deleting CloudFormation stacks. The hooks are very useful for tying different stacks together and generally supplementing what CloudFormation can do.

**Comment by kkesley-shine** [comment]
Another use case is to create a codebuild's build parameters in a json file. Currently, we have to manually create it after the stack is created. with this feature, we can automate the creation easily.

**Comment by andreimcristof** [comment]
+1. 
My use case is that I need to write a config file post-deployment, which contains the resulting  endpoints and ARN's and such.

**Comment by davidolmo** [comment]
+1
This is required in order to connect the created resources with some external site for CI/CD (for example bitbucket pipelines) and avoiding setting the names of all resources manually

**Comment by markus7811** [comment]
Another use case would be to use this with eks iam integration - patching k8s service accounts after the iam role with oidc connect is created (need to know the name of the role, so only can patch it after the stack was created). Would like to see something like this ...

**Comment by andreimcristof** [comment]
If I have a nodejs app that uses aws-sdk to populate a dynamodb table from a json, and I want to run this after the whole provisioning has completed, is this a case for this exact issue or can it be done in another way? Thanks,

**Comment by jogold** [comment]
@andreimcristof you could use a custom resource for this (
https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-cloudformation/README.md#custom-resources and
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/template-custom-resources.html).

**Comment by eladb** [comment]
I am working on some new high level mini-framework for custom resources that will make this really easy to implement.

**Comment by kferrone** [comment]
All I want to do is print a string based on some info in the stack after it has been deployed. Not even trying to do anything fancy here. Is there any reason why it can't just accept a callback or be a promise to handle? In the mean time, I could use `wait-for` npm package to wait till it's done as long as there is some clear trigger for it to watch and wait for. Any ideas how one could use `wait-for`?

**Comment by piyushwadhwani** [comment]
@eladb  Any update on your custom resource mini framework

**Comment by eladb** [comment]
Yes. It’s available under the @aws-cdk/custom-resources module. Should be possible to implement this quite easily now.

**Comment by kferrone** [comment]
So in the mean time, the easiest way to run custom scripts after deploy is to use the `poststart` script in `package.json` and make sure your start script in package.json is `cdk deploy`. Then simply run `npm start` and when it is done deploying it will run your poststart script. Since CDK uses the aws-sdk package you can pretty much do whatever you want afterward.

**Comment by tuusberg** [comment]
+1

**Comment by saksham** [comment]
+1
